include/oneapi/dpl/pstl/histogram_impl.h

```diff
@@ -77,6 +79,54 @@ histogram(_ExecutionPolicy&& exec, _RandomAccessIterator1 first, _RandomAccessIt
+// and required `<`, `<=`, `+`, `-`, and `/` to be defined between _ValueType and
+// std::iterator_traits<_RandomAccessIterator1>::value_type rather than enforcing they were the same type or convertible
+template <typename _ExecutionPolicy, typename _RandomAccessIterator1, typename _Size, typename _RandomAccessIterator2,
+          typename _ValueType>
+std::enable_if_t<
+    oneapi::dpl::execution::is_execution_policy_v<std::decay_t<_ExecutionPolicy>> &&
+        !std::is_convertible_v<_ValueType, typename std::iterator_traits<_RandomAccessIterator1>::value_type>,
+    _RandomAccessIterator2>
+histogram(_ExecutionPolicy&& exec, _RandomAccessIterator1 first, _RandomAccessIterator1 last, _Size num_bins,
+          _ValueType first_bin_min_val, _ValueType last_bin_max_val, _RandomAccessIterator2 histogram_first)
+{
+    const auto __dispatch_tag = oneapi::dpl::__internal::__select_backend(exec, first, histogram_first);
+
+    oneapi::dpl::__internal::__pattern_histogram(
+        __dispatch_tag, ::std::forward<_ExecutionPolicy>(exec), first, last, num_bins,
+        oneapi::dpl::__internal::__evenly_divided_binhash<_ValueType>(first_bin_min_val, last_bin_max_val, num_bins),
+        histogram_first);
+    return histogram_first + num_bins;
+}
+
+// This overload is provided to support an extension to the oneDPL specification to support the original implementation
+// of the histogram API, where if users explicitly-specify all template arguments, their arguments for bin boundary min
+// and max are convertible to the specified _ValueType, and the _ValueType is convertible to the value type of the
+// input iterator.
+template <typename _ExecutionPolicy, typename _RandomAccessIterator1, typename _Size, typename _RandomAccessIterator2,
+          typename _ValueType, typename _RealValueType1, typename _RealValueType2>
+std::enable_if_t<
+    oneapi::dpl::execution::is_execution_policy_v<std::decay_t<_ExecutionPolicy>> &&
+        std::is_convertible_v<_ValueType, typename std::iterator_traits<_RandomAccessIterator1>::value_type> &&
+        std::is_convertible_v<_RealValueType1, _ValueType> && std::is_convertible_v<_RealValueType2, _ValueType>,
+    _RandomAccessIterator2>
+histogram(_ExecutionPolicy&& exec, _RandomAccessIterator1 first, _RandomAccessIterator1 last, _Size num_bins,
+          _RealValueType1 first_bin_min_val, _RealValueType2 last_bin_max_val, _RandomAccessIterator2 histogram_first)
+{
+    const auto __dispatch_tag = oneapi::dpl::__internal::__select_backend(exec, first, histogram_first);
+
+    oneapi::dpl::__internal::__pattern_histogram(
+        __dispatch_tag, ::std::forward<_ExecutionPolicy>(exec), first, last, num_bins,
+        oneapi::dpl::__internal::__evenly_divided_binhash<_ValueType>(_ValueType{first_bin_min_val},
+                                                                      _ValueType{last_bin_max_val}, num_bins),
+        histogram_first);
+    return histogram_first + num_bins;
+}
```

Comment:
In this implementation, why not instead call the main specified overload, just converting the parameters to `_ValueType`?

Reply:
We could I think.  I was trying to follow the convention to not call external oneDPL APIs from within implementation, but this may be a case where an exception would be useful.

Reply:
After some testing, from my perspective this is not a good idea.

The two options are subtly different from each other.  As written currently in the PR, the functionality for this overload with explicit template arguments specified is exactly the same as it was before.  The user specifies that the bins be calculated from their chosen type, and we do so, using the operations (`+` `-` `\` `<` `<=`).

If we change to instead wrap and call the spec compliant signature, we do a conversion to the value type of the input iterator at the start, then internally operate in that type.  
With that change, we are doing `IteratorValueType{ExplicitValueType{argument}}` where the current implementation just does `ExplicitValueType{argument}`. If we are supporting these as an extension only, I think it makes sense to support them exactly as they were.  Both options work when matched with the correct corresponding call to the serial implementation in the tests, but I prefer it as it is written in the PR currently.

Reply:
OK, I don't mind enough to argue :)